File: exercism/python/sum-of-multiples/sum_of_multiples.py
"""
Sum of Multiples
"""


def sum_of_multiples(limit, multiples):
    """
    sum of all the unique multiples of particular numbers up to
    but not including that number

        arg limit: (int) limit number
        arg multiples: (list) list of factors
        return: (int) sum of multiples
    """

    # # simple
    # sum = 0
    # for index in range(limit):
    #   for number in multiples:
    #     if number:
    #         if index % number == 0:
    #             sum+=index
    #             break
    # return sum

    # # improve efficiency, using a set,
    # # range increment and sum.
    # values = set()
    # for number in multiples:
    #     if number:
    #         for index in range(number,limit,number):
    #                 values.add(index)
    # return sum(values)

    # the same but using set comprehension
    return sum({index for number in multiples if number
                for index in range(number, limit, number)})

File: exercism/go/sum-of-multiples/sum_of_multiples.go
package summultiples

func SumMultiples(limit int, divisors ...int) int {
	sum := 0

	// // simple
	// for index := 1; index < limit; index++ {
	// 	for _, number := range divisors {
	// 		if number > 0 && index%number == 0 {
	// 			sum += index
	// 			break
	// 		}
	// 	}
	// }

	// improvement, using a map for duplicates
	// and increment by number
	duplicates := make(map[int]bool)
	for _, number := range divisors {
		if number != 0 {
			for index := number; index < limit; index += number {
				if !duplicates[index] {
					sum += index
					duplicates[index] = true
				}
			}
		}
	}

	return sum
}
